geo/Geoserver.py

```diff
@@ -393,6 +393,121 @@ def create_workspace(self, workspace: str):
+        """
+        Create a new security rule for either all workspaces or a subset of them based on the provided pattern.
+
+        Parameters
+        ----------
+        workspacePattern : str
+            workspace pattern. Apply the security rule to either all workspaces (using *) or 
+            a subset of them where pattern matches (contains) the workspace name.
+        permission : str
+            [r|w|a] is a placeholder for the permission type. 
+            r read-only access, w write access and a indicates full (read and write) access.
+        role : str
+            rol name
+        """
+
+        if role is None:
+            raise Exception("You must provide a role name")
+
+        workspacesRawData = self.get_workspaces()
+
+        if workspacesRawData == None:
+            raise Exception("There is not workpaces available")
+
+        workspaceList = workspacesRawData['workspaces']['workspace']
+
+        if workspacePattern != '*':
+            workspaceList = [
+                workspace for workspace in workspaceList if workspacePattern in workspace['name']]
+
+        if len(workspaceList) == 0:
+            raise Exception("The pattern does not match any workspace")
+
+        for workspace in workspaceList:
+            res = self.create_update_layer_rule(
+                workspace['name'], '*', permission, role)
+            print(res)
+
+    def create_update_layer_rule(self, workspace: str, layer: str, permission: str, role: str):
+        """
+        Create or update a security rule for the specified layer(s) within a workspace
+
+        Parameters
+        ----------
+        workspace : str
+            workspace name
+        layer : str,
+            layer name or * for all layers within the selected workspace
+        permission : str
+            [r|w| ] is a placeholder for the permission type. 
+            r read-only access, w write access and a indicates full (read and write) access.
+        role : str
+            rol name
+
+        Example: curl -v -u admin:geoserver -XPOST http://localhost:8080/geoserver/rest/security/acl/layers -H "accept: application/json" -H "content-type: application/xml" -d "<rules><rule resource=\"WORKSPACE_NAME.*.r\">ROL_NAME</rule></rules>"
+        """
+        try:
+            url = "{}/rest/security/acl/layers".format(self.service_url)
+            data = "<rules><rule resource='{}.{}.{}'>{}</rule></rules>".format(
+                workspace, layer, permission, role)
+            headers = {"content-type": "application/xml"}
+            r = requests.post(
+                url, data, auth=(self.username, self.password), headers=headers
+            )
+
+            if r.status_code == 200:
+                return "{} Rule created! Workspace: {} | Layer: {} | Permission: {} | Role: {} ".format(r.status_code, workspace, layer, permission, role)
+
+            if r.status_code == 409:
+                return self.update_layer_rule(workspace, '*', permission, role)
```

Comment:
No, the rule is defined as a set of 4 variables: workspaces, layer, permission and role. It is possible to have 2 different valid rules without any conflicts:

- `"taxid-00025.*.r": "BIOTABLERO-FRONT",`
- `"taxid-00025.layer1.r": "BIOTABLERO-FRONT",`

So, in this case is necessary to send all layers `(*)` as we are defining rules grouped by workspace instead of layer by layer.

Reply:
Exactly what you are saying, you are creating a new rule instead of modifying the existing one.

Let's say that this rule already exists: `"taxid-00025.layer1.r": "BIOTABLERO-FRONT"`
And now there is another user (`NEW_USER`) that needs read access to `layer1` from `taxid-00025` workspace, the correct modification would be `"taxid-00025.layer1.r": "BIOTABLERO-FRONT, NEW_USER"`. What you are doing changing to all layers would end up with these two rules, giving more access to the new user than the desired one:

- `"taxid-00025.layer1.r": "BIOTABLERO-FRONT"`
- `"taxid-00025.*.r": "NEW_USER"`